### PR TITLE
Dave's part 

### PR DESCRIPTION
**Pull request by ruhan-dave**
# Multi-Task BERT for Customer Review Analysis

I implemented a multi-task learning approach with a pre-trained BERT model to analyze customer reviews. The model  predicts both the **sentiment score** (regression) and the **topic category** (classification) of individual sub-comments within reviews to provide a more nuanced analysis of customer feedback. 

## Features

*   **Data Loading and Preprocessing:**
    *   Loads pandas dataframe and expands into a larger set (sentences become subsentences). 
    *   Cleans text data (lowercasing, punctuation removal, etc.).
*   **Multi-Task BERT Model:**
    *   Uses a pre-trained BERT model (`bert-base-uncased` by default) as a backbone.
    *   Implements separate output layers ("heads") for sentiment regression and topic classification.
    *   Allows for fine-tuning the entire model (including the BERT backbone).
*   **Training and Evaluation:**
    *   Splits data into training, validation, and testing sets.
    *   Uses PyTorch and the `transformers` library for training.
    *   Calculates and reports:
        *   RMSE (Root Mean Squared Error) for sentiment prediction.
        *   Accuracy for topic classification.
    *   Includes model checkpointing (saves the best model based on validation loss).
*   **Clustering (Optional):**
    *   Includes functionality for clustering reviews using Bayesian Gaussian Mixture Models (BGMMs).
    *   Calculates cluster probabilities for each review.
    *   Provides tools for interpreting and labeling clusters (analyzing centroids and top TF-IDF terms).
*  **Sampling:** Includes ability to sample sentences from each cluster
* **Saving and Loading:**
    * Provides functions to save the trained model (state_dict) and tokenizer for later use.
    * Includes function to load saved model
*  **Detailed output:** Includes logging to standard output and printing results that can be used for monitoring.

## Requirements

*   Python 3.7+
*   PyTorch (tested with 1.13+, but should work with later versions)
*   Transformers (Hugging Face) (tested with 4.30+, but should work with later versions)
*   Pandas
*   scikit-learn
*   NumPy
*   Sentence Transformers (optional, if using sentence embeddings)

## Installation

```bash
pip install torch transformers pandas scikit-learn sentence-transformers
pip install polars #optional
pip install scikit-fuzzy #optional

### CODE REVIEW

**Comment by howl-anderson** [comment]
The code is clean and with enough comments. The README is readable to me. Greate job! Approved.
